Description: Usage of the zlib compression library provided by the system
Forwarded: http://sf.net/support/tracker.php?aid=2918870
Author: Thomas Gaugler <thomas@dadie.net>

Index: nsis-2.46/Source/czlib.h
===================================================================
--- nsis-2.46/Source/czlib.h	(revision 6015)
+++ nsis-2.46/Source/czlib.h	(working copy)
@@ -18,7 +18,19 @@
 #define __CZLIB_H__
 
 #include "compressor.h"
-#include "zlib/ZLIB.H"
+#ifdef _WIN32
+#  include "zlib/ZLIB.H"
+#  define deflateInit2(strm, level, method, windowBits, memLevel, strategy) \
+     deflateInit2_(strm, level, method, windowBits, memLevel, strategy, \
+     NULL, sizeof(z_stream))
+#  define streamInit(strm)
+#else
+#  include <zlib.h>
+#  define streamInit(strm) { \
+      strm->zalloc = (alloc_func)Z_NULL; \
+      strm->zfree = (free_func)Z_NULL; \
+      strm->opaque = (voidpf)Z_NULL; }
+#endif
 
 class CZlib : public ICompressor {
   public:
@@ -27,7 +39,9 @@
     int Init(int level, unsigned int dict_size) {
       stream = new z_stream;
       if (!stream) return Z_MEM_ERROR;
-      return deflateInit(stream, level);
+      streamInit(stream);
+      return deflateInit2(stream, level,
+         Z_DEFLATED, -MAX_WBITS, MAX_MEM_LEVEL, Z_DEFAULT_STRATEGY);
     }
 
     int End() {
Index: nsis-2.46/Source/SConscript
===================================================================
--- nsis-2.46/Source/SConscript	(revision 6015)
+++ nsis-2.46/Source/SConscript	(working copy)
@@ -66,6 +66,8 @@
 	# XXX will cause problems if makensis is cross compiled
 	# on freebsd, libversion.a exists and gives trouble if linked
 	libs += ['version']
+else:
+	libs += ['z']
 
 AddAvailableLibs(env, libs)
 
@@ -87,7 +89,9 @@
 
 ##### Compile makensis
 
-files = makensis_files + bzip2_files + lzma_files + zlib_files
+files = makensis_files + bzip2_files + lzma_files
+if env['PLATFORM'] == 'win32':
+	files += zlib_files
 
 makensis = env.Program(target, files)
 
Index: nsis-2.46/Source/Tests/SConscript
===================================================================
--- nsis-2.46/Source/Tests/SConscript	(revision 6015)
+++ nsis-2.46/Source/Tests/SConscript	(working copy)
@@ -1,3 +1,5 @@
+Import('env AddAvailableLibs')
+
 target = 'test'
 
 tests = Split("""
@@ -65,7 +67,9 @@
 	zlib/INFBLOCK.C
 """)
 
-required += zlib_files
+if env['PLATFORM'] == 'win32':
+	required += zlib_files
+
 required_exehead += zlib_exehead_files
 
 cppunitlibs = Split("""
@@ -87,8 +91,6 @@
 	winver.nsi
 """)
 
-Import('env AddAvailableLibs')
-
 # Test scripts
 env.TestScript(scripts)
 
@@ -97,6 +99,8 @@
 	# XXX will cause problems if tests are cross compiled
 	# on freebsd, libversion.a exists and gives trouble if linked
 	extralibs += ['version']
+else:
+	extralibs += ['z']
 
 AddAvailableLibs(env, extralibs)
 
